Contract development & interaction instructions for cw-to-do-list (#377)
InterWasm/docs: dev-academy/develop-smart-contract/02-develop.md

## Contract Config & State

Each instance of the To-Do List contract should have an `owner` with entry manipulation rights. In order to be able to check whether a function call is authorized or not, the address of the owner must be persisted on contract storage following the instantiation.

```rust
//state.rs
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct Config {
  pub owner: Addr,
}

pub const CONFIG: Item<Config> = Item::new("config");
//Item stores a single variable of a given type, identified by a string storage key.
```
A To-Do List consists of entries, each of which is identified by a unique `id`. The `id` of the latest entry on the list is stored as `ENTRY_SEQ`, an Item that holds an integer value. The value stored in `ENTRY_SEQ` is incremented every time a new entry is appended to the list.

The attributes `status` and `priority` are defined as enums to provide uniformity among different entries.

The list of entries as a whole is persisted as the `LIST`, a Map that holds a collection of `id` to `Entry` pairs.

```rust
//state.rs
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct Entry {

>     pub id: Uint64,

Reviewer: We can simply use `u64` for keys in storage-plus, best to reflect that in the docs

* The function finally saves the new entry to the `LIST` with the matching `id` and returns a `Response` with the relevant attributes. 

```rust
//contract.rs
pub fn create_new_entry(deps: DepsMut, info: MessageInfo, description: String, priority: Option<Priority>) -> Result<Response, ContractError> {
    let owner = CONFIG.load(deps.storage)?.owner;
    if info.sender != owner {
        return Err(ContractError::Unauthorized {});
    }
    let id = ENTRY_SEQ.update::<_, cosmwasm_std::StdError>(deps.storage, |id| {
        Ok(id.add(Uint64::new(1)))
    })?;
    let new_entry = Entry {
        id,
        description,
        priority: priority.unwrap_or(Priority::None),
        status: Status::ToDo
    };
    LIST.save(deps.storage, id.u64(), &new_entry)?;
    Ok(Response::new().add_attribute("method", "create_new_entry")
                      .add_attribute("new_entry_id: ", id))
}
```
The function `update_entry()` handles the update of an existing entry on the list. 

> * Before carrying on with the new update, the function checks if the message sender is the owner of the contract. If not, it returns an error and the update fails to be performed.

Reviewer: ```suggestion
* Before continuing with the new update, the function checks if the message sender is the owner of the contract. If not, it returns an error and the update fails to be performed.
```

    msg: ExecuteMsg,
) -> Result<Response, ContractError> {
    match msg {

>         ExecuteMsg::NewEntry {description, priority} => create_new_entry(deps, info, description, priority),

Reviewer: For function names, best to follow best practices. you should use `action_create...` naming convention

        ExecuteMsg::UpdateEntry {id, description, status, priority } => update_entry(deps, info, id, description, status, priority),
        ExecuteMsg::DeleteEntry {id} => delete_entry(deps, info, id)
    }
}
```

> The function `create_new_entry()` handles the creation of a new entry on the list. 

Reviewer: We can put these under new h3 headers
```
### Create New Entry

The function `create_new_entry()` handles the creation of a new entry on the list. 
```

> * Before creating the new entry, the function checks if the message sender is the owner of the contract. If not, it returns an error and the new entry creation fails to be performed.

Reviewer: For these steps better to use numbering, it indicates this is a ordered algorithm
```suggestion
1. Before creating the new entry, the function checks if the message sender is the owner of the contract. If not, it returns an error and the new entry creation fails to be performed.
```

    let id = ENTRY_SEQ.update::<_, cosmwasm_std::StdError>(deps.storage, |id| {
        Ok(id.add(1))
    })?;
    let new_entry = Entry {
        id,
        description,
        priority: priority.unwrap_or(Priority::None),
        status: Status::ToDo
    };
    LIST.save(deps.storage, id, &new_entry)?;
    Ok(Response::new().add_attribute("method", "create_new_entry")
        .add_attribute("new_entry_id", id.to_string()))
}
```
The function `update_entry()` handles the update of an existing entry on the list. 
* Before continuing with the new update, the function checks if the message sender is the owner of the contract. If not, it returns an error and the update fails to be performed.
* The entry with the matching `id` is loaded from the `LIST`.
* Sharing the same id, an updated version of the entry is defined with the received values for `description`, `status` and `priority`. These are optional parameters and if any one of them is not provided, the function defaults back to the corresponding value from the entry loaded.
* The function saves the updated entry to the `LIST` with the matching `id` and returns a `Response` with the relevant attributes. 

```rust
//contract.rs

> pub fn update_entry(deps: DepsMut, info: MessageInfo, id: u64, description: Option<String>, status: Option<Status>, priority: Option<Priority>) -> Result<Response, ContractError> {

Reviewer: Just a thought. Maybe you can explain the steps above in code comments? So the logic and implementation walks together

    };
    LIST.save(deps.storage, id.u64(), &new_entry)?;
    Ok(Response::new().add_attribute("method", "create_new_entry")

>                       .add_attribute("new_entry_id: ", id))

Reviewer: ```suggestion
                      .add_attribute("new_entry_id", id))
```
no need for `:`

    };
    LIST.save(deps.storage, id.u64(), &updated_entry)?;
    Ok(Response::new().add_attribute("method", "update_entry")

>                       .add_attribute("updated_entry_id: ", id))

Reviewer: ```suggestion
                      .add_attribute("updated_entry_id", id))
```

    LIST.remove(deps.storage, id.u64());
    Ok(Response::new().add_attribute("method", "delete_entry")

>                       .add_attribute("deleted_entry_id: ", id))

Reviewer: ```suggestion
                      .add_attribute("deleted_entry_id", id))
```
